routing/online_cross_fetcher.cpp

```diff
@@ -21,50 +21,67 @@ inline string LatLonToURLArgs(ms::LatLon const & point)
 
 namespace routing
 {
-bool ParseResponse(const string & serverResponse, vector<m2::PointD> & outPoints)
+bool ParseResponse(string const & serverResponse, vector<m2::PointD> & outPoints)
 {
   try
   {
     my::Json parser(serverResponse.c_str());
 
     json_t const * countries = json_object_get(parser.get(), "used_mwms");
-    size_t pointsCount = json_array_size(countries);
-    outPoints.reserve(pointsCount);
+    size_t const pointsCount = json_array_size(countries);
     for (size_t i = 0; i < pointsCount; ++i)
     {
       json_t * pointArray = json_array_get(countries, i);
       outPoints.push_back({json_number_value(json_array_get(pointArray, 0)),
                            json_number_value(json_array_get(pointArray, 1))});
     }
-    return !outPoints.empty();
+    return pointsCount > 0;
   }
-  catch (my::Json::Exception&)
+  catch (my::Json::Exception const & exception)
   {
+    LOG(LWARNING, ("Can't parse server response:", exception.what()));
+    LOG(LWARNING, ("Response body:", serverResponse));
     return false;
   }
-  return false;
 }
 
 string GenerateOnlineRequest(string const & serverURL, ms::LatLon const & startPoint,
                              ms::LatLon const & finalPoint)
 {
   return serverURL + "/mapsme?loc=" + LatLonToURLArgs(startPoint) + "&loc=" +
-         LatLonToURLArgs(finalPoint);
+    LatLonToURLArgs(finalPoint);
 }
 
-OnlineCrossFetcher::OnlineCrossFetcher(string const & serverURL, ms::LatLon const & startPoint,
-                                       ms::LatLon const & finalPoint)
-    : m_request(GenerateOnlineRequest(serverURL, startPoint, finalPoint))
+OnlineCrossFetcher::OnlineCrossFetcher(TCountryFileFn const & countryFileFn,
+                                       string const & serverURL, Checkpoints const & checkpoints)
+  : m_countryFileFn(countryFileFn), m_serverURL(serverURL), m_checkpoints(checkpoints)
 {
-  LOG(LINFO, ("Check mwms by URL: ", GenerateOnlineRequest(serverURL, startPoint, finalPoint)));
+  CHECK(m_countryFileFn, ());
 }
 
 void OnlineCrossFetcher::Do()
 {
   m_mwmPoints.clear();
-  if (m_request.RunHttpRequest() && m_request.ErrorCode() == 200 && !m_request.WasRedirected())
-    ParseResponse(m_request.ServerResponse(), m_mwmPoints);
-  else
-    LOG(LWARNING, ("Can't get OSRM server response. Code: ", m_request.ErrorCode()));
+
+  for (size_t i = 0; i < m_checkpoints.GetNumSubroutes(); ++i)
+  {
+    m2::PointD const & pointFrom = m_checkpoints.GetPoint(i);
+    m2::PointD const & pointTo = m_checkpoints.GetPoint(i + 1);
+
+    string const mwmFrom = m_countryFileFn(pointFrom);
+    string const mwmTo = m_countryFileFn(pointTo);
+    if (mwmFrom == mwmTo)
+      continue;
+
+    string const url = GenerateOnlineRequest(m_serverURL, MercatorBounds::ToLatLon(pointFrom),
+                                             MercatorBounds::ToLatLon(pointTo));
+    platform::HttpClient request(url);
```

Comment:
Do you plan to fix this? From my point of view, there'll be not so much changes on the client side and on the server side?